File: systems/plants/IKoptions.cpp
#include "RigidBodyManipulator.h"
#include "IKoptions.h"
using namespace std;
using namespace Eigen;

IKoptions::IKoptions(RigidBodyManipulator* robot)
{
  // It is important to make sure these default values are consistent with the MATLAB IKoptions
  this->setDefaultParams(robot);
}

IKoptions::IKoptions(const IKoptions &rhs)
{
  this->robot = rhs.robot;
  this->nq = rhs.nq;
  this->Q = rhs.Q;
  this->Qa = rhs.Qa;
  this->Qv = rhs.Qv;
  this->debug_mode = rhs.debug_mode;
  this->sequentialSeedFlag = rhs.sequentialSeedFlag;
  this->SNOPT_MajorFeasibilityTolerance = rhs.SNOPT_MajorFeasibilityTolerance;
  this->SNOPT_MajorIterationsLimit = rhs.SNOPT_MajorIterationsLimit;
  this->SNOPT_IterationsLimit = rhs.SNOPT_IterationsLimit;
  this->SNOPT_SuperbasicsLimit = rhs.SNOPT_SuperbasicsLimit;
  this->SNOPT_MajorOptimalityTolerance = rhs.SNOPT_MajorOptimalityTolerance;
  this->additional_tSamples = rhs.additional_tSamples;
  this->fixInitialState = rhs.fixInitialState;
  this->q0_lb = rhs.q0_lb;
  this->q0_ub = rhs.q0_ub;
  this->qd0_lb = rhs.qd0_lb;
  this->qd0_ub = rhs.qd0_ub;
  this->qdf_lb = rhs.qdf_lb;
  this->qdf_ub = rhs.qdf_ub;
}
IKoptions::~IKoptions()
{
}

void IKoptions::setDefaultParams(RigidBodyManipulator* robot)
{
  this->robot = robot;
  this->nq = this->robot->num_positions;
  this->Q = MatrixXd::Identity(this->nq,this->nq);
  this->Qa = 0.1*MatrixXd::Identity(this->nq,this->nq);
  this->Qv = MatrixXd::Zero(this->nq,this->nq);
  this->debug_mode = true;
  this->sequentialSeedFlag = false;
  this->SNOPT_MajorFeasibilityTolerance = 1E-6;
  this->SNOPT_MajorIterationsLimit = 200;
  this->SNOPT_IterationsLimit = 10000;
  this->SNOPT_SuperbasicsLimit = 2000;
  this->SNOPT_MajorOptimalityTolerance = 1E-4;
  this->additional_tSamples.resize(0);
  this->fixInitialState = true;
  this->q0_lb = this->robot->joint_limit_min;
  this->q0_ub = this->robot->joint_limit_max;
  this->qd0_ub = VectorXd::Zero(this->nq);
  this->qd0_lb = VectorXd::Zero(this->nq);
  this->qdf_ub = VectorXd::Zero(this->nq);
  this->qdf_lb = VectorXd::Zero(this->nq);
}

RigidBodyManipulator* IKoptions::getRobotPtr() const
{
  return this->robot;
}

void IKoptions::setQ(const MatrixXd &Q)
{
  if(Q.rows() != this->nq || Q.cols() != this->nq)
  {
    cerr<<"Q should be nq x nq matrix"<<endl;
  }
  this->Q = (Q+Q.transpose())/2;
  SelfAdjointEigenSolver<MatrixXd> eigensolver(this->Q);
  VectorXd ev= eigensolver.eigenvalues();
  for(int i = 0;i<this->nq;i++)
  {
    if(ev(i)<0)
    {
      cerr<<"Q is not positive semi-definite"<<endl;
    }
  }
}

void IKoptions::setQa(const MatrixXd &Qa)
{
  if(Qa.rows() != this->nq || Qa.cols() != this->nq)
  {
    cerr<<"Qa should be nq x nq matrix"<<endl;
  }
  this->Qa = (Qa+Qa.transpose())/2;
  SelfAdjointEigenSolver<MatrixXd> eigensolver(this->Qa);
  VectorXd ev= eigensolver.eigenvalues();
  for(int i = 0;i<this->nq;i++)
  {
    if(ev(i)<0)
    {
      cerr<<"Qa is not positive semi-definite"<<endl;
    }
  }
}

void IKoptions::setQv(const MatrixXd &Qv)
{
  if(Qv.rows() != this->nq || Qv.cols() != this->nq)
  {
    cerr<<"Qv should be nq x nq matrix"<<endl;
  }
  this->Qv = (Qv+Qv.transpose())/2;
  SelfAdjointEigenSolver<MatrixXd> eigensolver(this->Qv);
  VectorXd ev= eigensolver.eigenvalues();
  for(int i = 0;i<this->nq;i++)
  {
    if(ev(i)<0)
    {
      cerr<<"Qv is not positive semi-definite"<<endl;
    }
  }
}
void IKoptions::getQ(MatrixXd &Q) const
{
  Q = this->Q;
}

void IKoptions::getQa(MatrixXd &Qa) const
{
  Qa = this->Qa;
}

void IKoptions::getQv(MatrixXd &Qv) const
{
  Qv = this->Qv;
}

void IKoptions::setDebug(bool flag)
{
  this->debug_mode = flag;
}

bool IKoptions::getDebug() const
{
  return this->debug_mode;
}

void IKoptions::setSequentialSeedFlag(bool flag)
{
  this->sequentialSeedFlag = flag;
}

bool IKoptions::getSequentialSeedFlag() const
{
  return this->sequentialSeedFlag;
}

void IKoptions::setMajorOptimalityTolerance(double tol)
{
  if(tol<=0)
  {
    cerr<<"Major Optimality Tolerance must be positive"<<endl;
  }
  this->SNOPT_MajorOptimalityTolerance = tol;
}

double IKoptions::getMajorOptimalityTolerance() const
{
  return this->SNOPT_MajorOptimalityTolerance;
}

void IKoptions::setMajorFeasibilityTolerance(double tol)
{
  if(tol<=0)
  {
    cerr<<"Major Feasibility Tolerance must be positive"<<endl;
  }
  this->SNOPT_MajorFeasibilityTolerance = tol;
}

double IKoptions::getMajorFeasibilityTolerance() const
{
  return this->SNOPT_MajorFeasibilityTolerance;
}

void IKoptions::setSuperbasicsLimit(int limit)
{
  if(limit<=0)
  {
    cerr<<"Superbasics limit must be positive"<<endl;
  }
  this->SNOPT_SuperbasicsLimit = limit;
}

int IKoptions::getSuperbasicsLimit() const
{
  return this->SNOPT_SuperbasicsLimit;
}

void IKoptions::setMajorIterationsLimit(int limit)
{
  if(limit<=0)
  {
    cerr<<"Major iterations limit must be positive"<<endl;
  }
  this->SNOPT_MajorIterationsLimit = limit;
}

int IKoptions::getMajorIterationsLimit() const
{
  return this->SNOPT_MajorIterationsLimit;
}

void IKoptions::setIterationsLimit(int limit)
{
  if(limit<=0)
  {
    cerr<<"Iterations limit must be positive"<<endl;
  }
  this->SNOPT_IterationsLimit = limit;
}

int IKoptions::getIterationsLimit() const
{
  return this->SNOPT_IterationsLimit;
}

void IKoptions::setFixInitialState(bool flag)
{
  this->fixInitialState = flag;
}

bool IKoptions::getFixInitialState() const
{
  return this->fixInitialState;
}

void IKoptions::setq0(const VectorXd &lb, const VectorXd& ub)
{
  if(lb.rows() != this->nq || ub.rows() != this->nq)
  {
    cerr<<"q0_lb and q0_ub must be nq x 1 column vector"<<endl;
  }
  this->q0_lb = lb;
  this->q0_ub = ub;
  for(int i = 0;i<this->nq;i++)
  {
    if(this->q0_lb(i)>this->q0_ub(i))
    {
      cerr<<"q0_lb must be no larger than q0_ub"<<endl;
    }
    this->q0_lb(i) = this->q0_lb(i)>this->robot->joint_limit_min[i]? this->q0_lb(i): this->robot->joint_limit_min[i];
    this->q0_ub(i) = this->q0_ub(i)<this->robot->joint_limit_max[i]? this->q0_ub(i): this->robot->joint_limit_max[i];
  }
}

void IKoptions::getq0(VectorXd &lb, VectorXd &ub) const
{
  lb = this->q0_lb;
  ub = this->q0_ub;
}

void IKoptions::setqd0(const VectorXd &lb, const VectorXd& ub)
{
  if(lb.rows() != this->nq || ub.rows() != this->nq)
  {
    cerr<<"qd0_lb and qd0_ub must be nq x 1 column vector"<<endl;
  }
  for(int i = 0;i<this->nq;i++)
  {
    if(lb(i)>ub(i))
    {
      cerr<<"qd0_lb must be no larger than qd0_ub"<<endl;
    }
  }
  this->qd0_lb = lb;
  this->qd0_ub = ub;
}

void IKoptions::getqd0(VectorXd &lb, VectorXd &ub) const
{
  lb = this->qd0_lb;
  ub = this->qd0_ub;
}

void IKoptions::setqdf(const VectorXd &lb, const VectorXd& ub)
{
  if(lb.rows() != this->nq || ub.rows() != this->nq)
  {
    cerr<<"qdf_lb and qdf_ub must be nq x 1 column vector"<<endl;
  }
  for(int i = 0;i<this->nq;i++)
  {
    if(lb(i)>ub(i))
    {
      cerr<<"qdf_lb must be no larger than qdf_ub"<<endl;
    }
  }
  this->qdf_lb = lb;
  this->qdf_ub = ub;
}

void IKoptions::getqdf(VectorXd &lb, VectorXd &ub) const
{
  lb = this->qdf_lb;
  ub = this->qdf_ub;
}

void IKoptions::setAdditionaltSamples(const RowVectorXd &t_samples)
{
  if(t_samples.size()>0)
  {
    set<double> unique_sort_t(t_samples.data(),t_samples.data()+t_samples.size());
    this->additional_tSamples.resize(unique_sort_t.size());
    int t_idx = 0;
    for(auto it = unique_sort_t.begin();it != unique_sort_t.end();it++)
    {
      this->additional_tSamples(t_idx) = *it;
      t_idx++;
    }
  }
  else
  {
    this->additional_tSamples.resize(0);
  }
}

void IKoptions::getAdditionaltSamples(RowVectorXd &t_samples) const
{
  t_samples = this->additional_tSamples;
}

void IKoptions::updateRobot(RigidBodyManipulator* new_robot)
{
  this->robot = new_robot;
  int nq_cache = this->nq;
  this->nq = this->robot->num_positions;
  if(nq_cache != nq)
  {
    this->setDefaultParams(new_robot);
  }
}

File: systems/plants/IKoptions.m
classdef IKoptions
% @param Q                  --the cost matrix for posture deviation
% @param use_mex,           -- A boolean flat,set to true if run with mex. default is
%                              false
% @param debug_mode         -- IK would query the name of each constraint
%
% The following fields is used in inverseKinPointwise
% @param sequentialSeedFlag -- A boolean variable, set to false if the seed of
%                             inverseKinPointwise is q_seed. If set to true,
%                             inverseKinPointwise would use the solution in the
%                             previous time point as seed (if it is feasible).
%
% All subsequent fields will be used in inverseKinTraj only
% @param Qa                 -- The cost matrix for acceleration
% @param Qv                 -- The cost matrix for velocity
% @param additional_tSamples  -- Additional time samples (apart from knot points) in inverseKinTraj to check
%                                constraints
% @param fixInitialState    -- A boolean flag, set to true if the posture q and
%                              velocity qd at time t(1) is fixed. q(1) = q_seed_traj.eval(t(1))
%                              qdot(1) = (ikpotions.qd0_lb+ikoptions.qd0_ub)/2
% @param q0_lb              -- The initial posture upperbound. Would be used when
%                              fixInitialState = false;
% @param q0_ub              -- The initial posture lowerbound, Would be used when
%                              fixInitialState = false;
% @param qd0_ub             -- The upperbound of the initial velocity
% @param qd0_lb             -- The lowerbound of the initial velocity
% @param qdf_ub             -- The upper bound of the final velocity
% @param qdf_lb             -- The lower bound of the final velocity
  properties(SetAccess=protected)
    robot
    nq
    Q
    Qa
    Qv
    use_mex
    debug_mode
    sequentialSeedFlag
    SNOPT_MajorFeasibilityTolerance
    SNOPT_MajorIterationsLimit
    SNOPT_IterationsLimit
    SNOPT_SuperbasicsLimit
    SNOPT_MajorOptimalityTolerance
    additional_tSamples
    fixInitialState
    q0_lb
    q0_ub
    qd0_lb
    qd0_ub
    qdf_ub
    qdf_lb
    mex_ptr = nullPointer();
  end

  methods(Access = protected)
    function obj = setDefaultParams(obj,robot)
      obj.robot = robot;
      obj.nq = obj.robot.getNumPositions();
      obj.Q = eye(obj.nq);
      obj.Qa = 0.1*eye(obj.nq);
      obj.Qv = zeros(obj.nq);
      obj.use_mex = logical(exist('inverseKinmex','file'));
      obj.debug_mode = true;
      obj.sequentialSeedFlag = false;
      obj.SNOPT_MajorFeasibilityTolerance = 1e-6;
      obj.SNOPT_MajorIterationsLimit = 200;
      obj.SNOPT_IterationsLimit = 10000;
      obj.SNOPT_SuperbasicsLimit = 2000;
      obj.SNOPT_MajorOptimalityTolerance = 1e-4;
      obj.additional_tSamples = [];
      obj.fixInitialState = true;
      [obj.q0_lb,obj.q0_ub] = obj.robot.getJointLimits();
      obj.qd0_ub = zeros(obj.nq,1);
      obj.qd0_lb = zeros(obj.nq,1);
      obj.qdf_ub = zeros(obj.nq,1);
      obj.qdf_lb = zeros(obj.nq,1);
    end
  end

  methods
    function obj = IKoptions(robot)
      obj = setDefaultParams(obj,robot);
      if robot.getMexModelPtr~=0 && exist('IKoptionsmex','file')
        ptr = IKoptionsmex(1,robot.getMexModelPtr);
        obj.mex_ptr = ptr;
      end
    end

    function obj = setQ(obj,Q)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'Q',Q);
      end
      typecheck(Q,'double');
      sizecheck(Q,[obj.nq,obj.nq]);
      Q = (Q+Q')/2;
      if(any(abs(eig(Q))<0))
        error('IKoptions:Q must be positive semidefinite');
      end
      obj.Q = Q;
    end

    function obj = setQa(obj,Qa)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'Qa',Qa);
      end
      typecheck(Qa,'double');
      sizecheck(Qa,[obj.nq,obj.nq]);
      Qa = (Qa+Qa')/2;
      if(any(abs(eig(Qa))<0))
        error('IKoptions:Qa must be positive semidefinite');
      end
      obj.Qa = Qa;
    end

    function obj = setQv(obj,Qv)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'Qv',Qv);
      end
      typecheck(Qv,'double');
      sizecheck(Qv,[obj.nq,obj.nq]);
      Qv = (Qv+Qv')/2;
      if(any(abs(eig(Qv))<0))
        error('IKoptions:Qv must be positive semidefinite');
      end
      obj.Qv = Qv;
    end

    function obj = setMex(obj,flag)
      sizecheck(flag,[1,1]);
      if(isa(flag,'logical'))
        obj.use_mex = flag;
      elseif(isa(flag,'double'))
        obj.use_mex = logical(flag);
      else
        error('IKoptions: Unsupported flag type');
      end
      if obj.use_mex && obj.mex_ptr==0
        warning('Drake:IKoptions:MexDisabled','You asked to enable mex, but mex support is currently disabled (because the IK mex has not been built)');
        obj.use_mex = false;
      end
    end

    function obj = setDebug(obj,flag)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'debug',flag);
      end
      sizecheck(flag,[1,1]);
      if(isa(flag,'logical'))
        obj.debug_mode = flag;
      elseif(isa(flag,'double'))
        obj.debug_mode = logical(flag);
      else
        error('IKoptions: Unsupported flag type');
      end
    end

    function obj = setSequentialSeedFlag(obj,flag)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'sequentialSeedFlag',flag);
      end
      sizecheck(flag,[1,1]);
      if(isa(flag,'logical'))
        obj.sequentialSeedFlag = flag;
      elseif(isa(flag,'double'))
        obj.sequentialSeedFlag = logical(flag);
      else
        error('Drake:IKoptions: flag must be a logical scalar');
      end
    end

    function obj = setMajorOptimalityTolerance(obj,tol)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'majorOptimalityTolerance',tol);
      end
      sizecheck(tol,[1,1]);
      typecheck(tol,'double');
      if(tol<=0)
        error('Drake:IKoptions: tol must be positive');
      end
      obj.SNOPT_MajorOptimalityTolerance = tol;
    end

    function obj = setMajorFeasibilityTolerance(obj,tol)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'majorFeasibilityTolerance',tol);
      end
      sizecheck(tol,[1,1]);
      typecheck(tol,'double');
      if(tol<=0)
        error('Drake:IKoptions: tol must be positive');
      end
      obj.SNOPT_MajorFeasibilityTolerance = tol;
    end

    function obj = setSuperbasicsLimit(obj,limit)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'superbasicsLimit',limit);
      end
      sizecheck(limit,[1,1]);
      typecheck(limit,'double');
      if(limit<1)
        error('Drake:IKoptions: limit must be positive integers');
      end
      obj.SNOPT_SuperbasicsLimit = floor(limit);
    end

    function obj = setMajorIterationsLimit(obj,limit)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'majorIterationsLimit',limit);
      end
      sizecheck(limit,[1,1]);
      typecheck(limit,'double');
      if(limit<1)
        error('Drake:IKoptions: limit must be positive integers');
      end
      obj.SNOPT_MajorIterationsLimit = floor(limit);
    end

    function obj = setIterationsLimit(obj,limit)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'iterationsLimit',limit);
      end
      sizecheck(limit,[1,1]);
      typecheck(limit,'double');
      if(limit<1)
        error('Drake:IKoptions: limit must be positive integers');
      end
      obj.SNOPT_IterationsLimit = floor(limit);
    end

    function obj = setFixInitialState(obj,flag)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'fixInitialState',flag);
      end
      sizecheck(flag,[1,1]);
      if(isa(flag,'logical'))
        obj.fixInitialState = flag;
      elseif(isa(flag,'double'))
        obj.fixInitialState = logical(flag);
      else
        error('IKoptions: Unsupported flag type');
      end
    end

    function obj = setq0(obj,lb,ub)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'q0',lb,ub);
      end
      typecheck(lb,'double');
      typecheck(ub,'double');
      sizecheck(lb,[obj.nq,1]);
      sizecheck(ub,[obj.nq,1]);
      if(any(lb>ub))
        error('IKoptions:setq0: lb must be no larger than ub');
      end
      [lb_tmp,ub_tmp] = obj.robot.getJointLimits();
      obj.q0_lb = max([lb lb_tmp],[],2);
      obj.q0_ub = min([ub ub_tmp],[],2);
    end

    function obj = setqd0(obj,lb,ub)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'qd0',lb,ub);
      end
      typecheck(lb,'double');
      typecheck(ub,'double');
      sizecheck(lb,[obj.nq,1]);
      sizecheck(ub,[obj.nq,1]);
      if(any(lb>ub))
        error('IKoptions:setqd0: lb must be no larger than ub');
      end
      obj.qd0_lb = lb;
      obj.qd0_ub = ub;
    end


    function obj = setqdf(obj,lb,ub)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'qdf',lb,ub);
      end
      typecheck(lb,'double');
      typecheck(ub,'double');
      sizecheck(lb,[obj.nq,1]);
      sizecheck(ub,[obj.nq,1]);
      if(any(lb>ub))
        error('IKoptions:setqdf: lb must be no larger than ub');
      end
      obj.qdf_lb = lb;
      obj.qdf_ub = ub;
    end

    function obj = setAdditionaltSamples(obj,t_samples)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'additionaltSamples',t_samples);
      end
      typecheck(t_samples,'double');
      if(~isempty(t_samples))
        sizecheck(t_samples,[1,nan]);
        obj.additional_tSamples = unique(t_samples);
      else
        obj.additional_tSamples = [];
      end
    end

    function obj = updateRobot(obj,robot)
      if obj.mex_ptr ~= 0
        obj.mex_ptr = IKoptionsmex(3,obj.mex_ptr,'robot',robot.getMexModelPtr);
      end
      obj.robot = robot;
      nq_cache = obj.nq;
      obj.nq = obj.robot.getNumPositions();
      if(obj.nq ~= nq_cache)
        obj = setDefaultParams(obj,robot);
      end
    end
  end
end
